# Welcome to Flow!

In [1]:
import pandas as pd
import numpy as np
import matplotlib as plt

import flow

data = flow.configure_data()

In [5]:
df.to_csv(r'C:\Users\mong275\Local Files\Sankey_Notebook_Development\SankeyFiles\discharge.csv')

In [2]:
df = flow.calc_wastewater_exports(data=data)
df

FIPS State             County  total_wastewater_discharge  \
0     01001    AL     Autauga County                    0.206327   
1     01003    AL     Baldwin County                    0.465987   
2     01005    AL     Barbour County                    0.123504   
3     01007    AL        Bibb County                    0.083538   
4     01009    AL      Blount County                    0.374497   
...     ...   ...                ...                         ...   
3219  72151    PR  Yabucoa Municipio                    0.077568   
3220  72153    PR    Yauco Municipio                    0.031176   
3221  78010    VI   St. Croix County                    0.000000   
3222  78020    VI    St. John County                    0.000000   
3223  78030    VI  St. Thomas County                    0.000000   

      residential_wastewater_discharge_ratio  \
0                                  11.467667   
1                                  17.510736   
2                                  10.320666   
3                                  15.888889   
4                                   5.444987   
...                                      ...   
3219                               25.252119   
3220                               55.903846   
3221                                     inf   
3222                                     inf   
3223                                     inf   

      commercial_wastewater_discharge_ratio  \
0                                  1.317007   
1                                  4.839827   
2                                  4.129745   
3                                  2.412698   
4                                  0.650000   
...                                     ...   
3219                               8.540541   
3220                              18.000000   
3221                                    inf   
3222                                    inf   
3223                                    inf   

      industrial_wastewater_discharge_ratio  wastewater_export_mgd  \
0                                       1.0                    0.0   
1                                       1.0                    0.0   
2                                       1.0                    0.0   
3                                       1.0                    0.0   
4                                       1.0                    0.0   
...                                     ...                    ...   
3219                                    1.0                    0.0   
3220                                    1.0                    0.0   
3221                                    NaN                    0.0   
3222                                    NaN                    0.0   
3223                                    NaN                    0.0   

      wastewater_import_mgd  
0                    -2.863  
1                   -13.606  
2                    -1.808  
3                    -0.734  
4                    -1.008  
...                     ...  
3219                 -0.990  
3220                 -1.610  
3221                 -4.630  
3222                  0.000  
3223                  0.000  

[3224 rows x 9 columns]

In [9]:
name_list = ['electricity to rejected energy', 'energy discharge from sectoral use', 'energy in wastewater',
                      'energy in agriculture', 'energy in public water supply', 'energy production exports',
                     'water discharge from sectoral use']
name_list.index('electricity to rejected energy')

0

In [7]:
water_discharge_dict = {'surfacewater': {'fresh': {'wastewater': 0, 'ocean': 0, 'surface': 1},
                                        'saline': {'wastewater': 0, 'ocean': 1, 'surface': 0}},
                        'groundwater': {'fresh': {'wastewater': 0, 'ocean': 0, 'surface': 1},
                                        'saline': {'wastewater': 0, 'ocean': 1, 'surface': 0}},
                                'pws': {'fresh': {'wastewater': 1, 'ocean': 0, 'surface': 0}}
                             }

In [11]:
sector_consumption_dict = {'residential': {'groundwater': {'saline': 0, 'fresh': .3},
                                                   'surfacewater': {'saline': 0, 'fresh': .3},
                                                   'pws': {'fresh': .3}},
                            'commercial': {'groundwater': {'saline': 0, 'fresh': .15},
                                           'surfacewater': {'saline': 0, 'fresh': .15},
                                           'pws': {'fresh': .15}},
                            'industrial':{'groundwater': {'saline': .003, 'fresh': .15},
                                          'surfacewater': {'saline': .003, 'fresh': .15},
                                          'pws': {'fresh': .15}},
                            'mining': {'groundwater': {'saline': .03, 'fresh': .15},
                                       'surfacewater': {'saline': .03, 'fresh': .15}},
                            'crop_irrigation': {'groundwater': {'saline': 0, 'fresh': .3},
                                                'surfacewater': {'saline': 0, 'fresh': .3}},
                            'livestock':  {'groundwater': {'saline': 0, 'fresh': .87},
                                           'surfacewater': {'saline': 0, 'fresh': .87}},
                            'aquaculture': {'groundwater': {'saline': 0, 'fresh': .5},
                                            'surfacewater': {'saline': 0, 'fresh': .5}}}

In [16]:
df = data.copy()

In [ ]:
saline_groundwater_industrial_mgd                                                                   42.9
saline_surfacewater_industrial_mgd                                                                742.92

In [2]:
df = flow.calc_sectoral_use_water_discharge(data=data, total=True)

In [8]:
x=16987.779934 +  6225.87913 + 4263.081345
x

27476.740408999998

In [9]:
df1=flow.calc_sectoral_use_water_discharge(data=data, total=True)
df1 = df1[['FIPS', 'residential_wastewater_discharge_mgd', 'commercial_wastewater_discharge_mgd', 'industrial_wastewater_discharge_mgd']]

In [14]:
df2 = flow.prep_wastewater_data()
df2 = df2[['FIPS', 'State', 'County', 'municipal_wastewater_mgd', 'infiltration_wastewater_mgd']]

In [23]:
df3 = pd.merge(df2, df1, how='left')
df3['sector_total'] = (df3['residential_wastewater_discharge_mgd'] 
                                                 + df3['commercial_wastewater_discharge_mgd'] 
                                                 + df3['industrial_wastewater_discharge_mgd'])
df3['diff'] = df3['municipal_wastewater_mgd'] - df3['sector_total']
df3.sum()

FIPS                                    0100101003010050100701009010110101301015010170...
State                                   ALALALALALALALALALALALALALALALALALALALALALALAL...
County                                  Autauga CountyBaldwin CountyBarbour CountyBibb...
municipal_wastewater_mgd                                                        32984.606
infiltration_wastewater_mgd                                                       806.044
residential_wastewater_discharge_mgd                                         16987.779934
commercial_wastewater_discharge_mgd                                            6225.87913
industrial_wastewater_discharge_mgd                                           4263.081345
sector_total                                                                 27476.740409
diff                                                                          5507.865591
dtype: object

In [38]:
df = pd.DataFrame()
df_sector_discharge = flow.calc_sectoral_use_water_discharge(data=data, total=True)
df_sector_discharge = df_sector_discharge[['FIPS', 'State', 'County',
                                              'residential_wastewater_discharge_mgd',
                                              'commercial_wastewater_discharge_mgd',
                                              'industrial_wastewater_discharge_mgd']]



df['total_wastewater_discharge'] = df_sector_discharge[['residential_wastewater_discharge_mgd',
                                              'commercial_wastewater_discharge_mgd',
                                              'industrial_wastewater_discharge_mgd']].sum(axis=1)

df['wastewater_exports_mgd'] = data['municipal_wastewater_mgd'] - df['total_wastewater_discharge']
df

total_wastewater_discharge  wastewater_exports_mgd
0                       2.844153                0.018847
1                      10.881048                2.724952
2                       1.908191               -0.100191
3                       1.612425               -0.878425
4                       2.657049               -1.649049
5                       1.621644               -0.421644
6                       1.440894               -0.055894
7                      10.265992                0.063008
8                       2.082348                0.727652
9                       0.790208               -0.003208
10                      1.540936                0.294064
11                      0.174360                0.043640
12                      1.524846               -0.299846
13                      0.424668                0.300332
14                      0.503235               -0.143235
15                      3.565649               -0.927649
16                      5.313808               -0.527808
17                      0.557471               -0.230471
18                      0.312941               -0.202941
19                      2.765359                0.212641
20                      0.726434               -0.212434
21                      6.008955               -2.930955
22                      2.461440               -0.667440
23                      2.401796                1.810204
24                      3.994264               -1.391264
25                      4.909064                1.230936
26                      2.206111                0.945889
27                      9.716330                0.250670
28                      0.680897                0.479103
29                      1.713657                0.525343
30                      0.646463                0.302537
31                      0.225882               -0.058882
32                      1.306252               -0.228252
33                      0.790268               -0.212268
34                      6.172490                5.573510
35                      3.949454                0.065546
36                     76.937716                6.381284
37                      0.648093                0.091907
38                      9.686506               -1.746506
39                      1.365609               -0.130609
40                     11.655312                0.304688
41                      4.539268                0.665732
42                      0.547177               -0.327177
43                      1.165165                1.021835
44                     39.483290               -2.174290
45                      0.746571                1.415429
46                      1.175063                0.577937
47                     14.597264               -0.087264
48                     95.768506              -87.442506
49                      2.163239               -1.169239
50                     23.470900                3.260100
51                     27.847715               -7.898715
52                      0.363722                0.714278
53                      1.342749                0.691251
54                      3.422612               -0.041612
55                      0.206496                1.930504
56                      7.818545               -4.898545
57                      5.128381               -1.855381
58                     11.231774               -2.306774
59                      0.806369               -0.117369
60                      6.873630                3.278370
61                      5.425689                1.106311
62                     16.330123                2.957877
63                      3.829626                0.231374
64                      0.443475               -0.302475
65                      0.351416                0.044584
66                      1.403528               -1.343528
67                      0.651387               -0.651387
68                      0.618947               -0.135947
69              

In [ ]:
df1 = flow.calc_electricity_rejected_energy(data=data)
df2 = flow.calc_sectoral_use_energy_discharge(data=data)
df3 = flow.calc_energy_wastewater(data=data)
df4 = flow.calc_energy_agriculture(data=data)
df5 = flow.calc_energy_pws(data=data)
df6 = flow.calc_energy_production_exports(data=data)
df7 = flow.calc_sectoral_use_water_discharge(data=data)
df_list = [df1, df2, df3, df4, df5, df6, df7]
df_list

In [30]:
df = flow.calc_sectoral_use_water_discharge(data=data)
for col in df.columns:
    print(col)

FIPS
State
County
residential_saline_groundwater_consumption_mgd
residential_fresh_groundwater_consumption_mgd
residential_saline_surfacewater_consumption_mgd
residential_fresh_surfacewater_consumption_mgd
residential_fresh_pws_consumption_mgd
commercial_saline_groundwater_consumption_mgd
commercial_fresh_groundwater_consumption_mgd
commercial_saline_surfacewater_consumption_mgd
commercial_fresh_surfacewater_consumption_mgd
commercial_fresh_pws_consumption_mgd
industrial_saline_groundwater_consumption_mgd
industrial_fresh_groundwater_consumption_mgd
industrial_saline_surfacewater_consumption_mgd
industrial_fresh_surfacewater_consumption_mgd
industrial_fresh_pws_consumption_mgd
mining_saline_groundwater_consumption_mgd
mining_fresh_groundwater_consumption_mgd
mining_saline_surfacewater_consumption_mgd
mining_fresh_surfacewater_consumption_mgd
crop_irrigation_saline_groundwater_consumption_mgd
crop_irrigation_fresh_groundwater_consumption_mgd
crop_irrigation_saline_surfacewater_consumpti

In [31]:
df2 = df[['residential_saline_groundwater_consumption_mgd',
'residential_fresh_groundwater_consumption_mgd',
'residential_saline_surfacewater_consumption_mgd',
'residential_fresh_surfacewater_consumption_mgd',
'residential_fresh_pws_consumption_mgd',]].sum()
#df[['residential_fresh_pws_consumption_mgd']].sum()
df2

residential_saline_groundwater_consumption_mgd        0.000000
residential_fresh_groundwater_consumption_mgd       733.371590
residential_saline_surfacewater_consumption_mgd       0.000000
residential_fresh_surfacewater_consumption_mgd       17.637036
residential_fresh_pws_consumption_mgd              6309.030066
dtype: float64

In [32]:
x = 733+17+6309
x

7059

In [33]:
df = flow.calc_sectoral_use_water_discharge(data=data, total=True)
df.sum()

FIPS                                        0100101003010050100701009010110101301015010170...
State                                       ALALALALALALALALALALALALALALALALALALALALALALAL...
County                                      Autauga CountyBaldwin CountyBarbour CountyBibb...
residential_consumption_mgd                                                       7060.038692
commercial_consumption_mgd                                                                  0
industrial_consumption_mgd                                                        2071.504498
mining_consumption_mgd                                                             547.582352
crop_irrigation_consumption_mgd                                                     35146.119
livestock_consumption_mgd                                                         1796.333466
aquaculture_consumption_mgd                                                       2443.586535
residential_wastewater_discharge_mgd                        

In [3]:
df = flow.prep_consumption_fraction()
df.sum()

FIPS                                                   0100101003010050100701009010110101301015010170...
State                                                  ALALALALALALALALALALALALALALALALALALALALALALAL...
County                                                 Autauga CountyBaldwin CountyBarbour CountyBibb...
residential_fresh_surfacewater_consumption_fraction                                           983.056848
commercial_fresh_surfacewater_consumption_fraction                                            515.951231
industrial_fresh_surfacewater_consumption_fraction                                            485.905442
industrial_saline_surfacewater_consumption_fraction                                              9.68174
mining_fresh_surfacewater_consumption_fraction                                                 511.57624
mining_saline_surfacewater_consumption_fraction                                               106.173518
livestock_fresh_surfacewater_consumption_fraction      

In [21]:
df = flow.prep_water_use_2015(variables=['County','fresh_surfacewater_residential_mgd', 
                                         'fresh_groundwater_residential_mgd', 'fresh_pws_residential_mgd']).sum(axis=0)
df

County                                Autauga CountyBaldwin CountyBarbour CountyBibb...
fresh_surfacewater_residential_mgd                                                 49.1
fresh_groundwater_residential_mgd                                               3210.93
fresh_pws_residential_mgd                                                      23293.89
dtype: object

In [3]:
df = flow.calc_energy_wastewater(data=data, total=True)
#df = df[df.State == "TX"]
for col in df.columns:
    print(col)

FIPS
State
County
electricity_wastewater_bbtu
wastewater_rejected_energy_bbtu
wastewater_energy_services_bbtu


In [15]:
total_df

ace
0     01001
1     01003
2     01005
3     01007
4     01009
...     ...
3218  72151
3219  72153
3220  78010
3221  78020
3222  78030

[3223 rows x 1 columns]

In [46]:
#test_dict.update({'test_zero':x})
test_dict
rejected_energy_total = pd.DataFrame.from_dict(test_dict, orient='index').transpose().sum(axis=1)
rejected_energy_total



2524     0.045767
2525     1.866417
2526     0.000000
2527     0.000000
2528     0.000000
          ...    
2773     0.024313
2774    11.966520
2775     0.000000
2776     0.001430
2777     3.983119
Length: 254, dtype: float64

In [ ]:
x = np.zeros((len(df), 1))
test_dict.update({'test_zero': x})
test_dict['test_zero']

In [5]:
# Move this to top

    # calculate rejected energy and energy services for each fuel, agriculture, water, and pumping type
    for water_type in water_types_list:
        for pumping_type in pumping_type_dict:
            for agriculture_type in agriculture_type_list:
                for fuel_type in fuel_type_dict:
                    fuel_type_efficiency = 'agriculture_' + fuel_type + "_efficiency_pct"
                    if f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu' in energy_value_dict:
                        rejected_energy_name = f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_rejected_energy_bbtu'
                        energy_services_name = f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_energy_services_bbtu'
                        if fuel_type_efficiency in df.columns:
                            #df[f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_rejected_energy_bbtu'] = (1 - df[fuel_type_efficiency]) \
                            #                                                                                         * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']
                            #df[f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_energy_services_bbtu'] = (df[fuel_type_efficiency]) \
                            #                                                                                         * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']

                            rejected_energy_value = (1 - df[fuel_type_efficiency]) * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']
                            energy_services_value = (df[fuel_type_efficiency]) * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']

                            rejected_energy_dict.update({rejected_energy_name: rejected_energy_value})
                            energy_services_dict.update({energy_services_name: energy_services_value})

                        # if there's no fuel efficiency column, use defaults
                        else:
                            #df[f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_rejected_energy_bbtu'] = (1 -fuel_type_dict[fuel_type]) \
                            #                                                                             * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']
                            #df[f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_energy_services_bbtu'] = (fuel_type_dict[fuel_type]) \
                            #                                                                             * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']

                            rejected_energy_value = (1 -fuel_type_dict[fuel_type]) * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']
                            energy_services_value = (fuel_type_dict[fuel_type]) * df[f'{fuel_type}_{agriculture_type}_{water_type}_{pumping_type}_bbtu']

                            rejected_energy_dict.update({rejected_energy_name: rejected_energy_value})
                            energy_services_dict.update({energy_services_name: energy_services_value})

                        #df[f'{agriculture_type}_total_rejected_energy_bbtu'] = df[f'{agriculture_type}_total_rejected_energy_bbtu'] + df[f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_rejected_energy_bbtu']
                        #df[f'{agriculture_type}_total_energy_services_bbtu'] = df[f'{agriculture_type}_total_energy_services_bbtu'] + df[ f'{agriculture_type}_{water_type}_{pumping_type}_{fuel_type}_energy_services_bbtu']
                    else:
                        pass

pumping
treatment
distribution
pumping
treatment
distribution
pumping
treatment
distribution
pumping
treatment
distribution


Empty DataFrame
Columns: [pws_groundwater_total_energy_bbtu, pws_groundwater_total_energy_services_bbtu, pws_groundwater_total_rejected_energy_bbtu, pws_surface_water_total_energy_bbtu, pws_surface_water_total_energy_services_bbtu, pws_surface_water_total_rejected_energy_bbtu]
Index: []

In [7]:
df= df[df.State == "CA"]
df = df['electricity_irrigation_ibt_rejected_energy_bbtu']
df.sum()

9849.530549071445

In [2]:
df = flow.calc_energy_agriculture(data=data, total=True)
df

FIPS State             County  crop_irrigation_total_energy_bbtu  \
0     01001    AL     Autauga County                           0.016044   
1     01003    AL     Baldwin County                           0.237871   
2     01005    AL     Barbour County                           0.006298   
3     01007    AL        Bibb County                           0.000143   
4     01009    AL      Blount County                           0.001234   
...     ...   ...                ...                                ...   
3219  72151    PR  Yabucoa Municipio                           0.000000   
3220  72153    PR    Yauco Municipio                           0.000000   
3221  78010    VI   St. Croix County                           0.000000   
3222  78020    VI    St. John County                           0.000000   
3223  78030    VI  St. Thomas County                           0.000000   

      crop_irrigation_total_rejected_energy_bbtu  \
0                                       0.005616   
1                                       0.083255   
2                                       0.002204   
3                                       0.000050   
4                                       0.000432   
...                                          ...   
3219                                    0.000000   
3220                                    0.000000   
3221                                    0.000000   
3222                                    0.000000   
3223                                    0.000000   

      crop_irrigation_total_energy_services_bbtu  \
0                                       0.010429   
1                                       0.154616   
2                                       0.004094   
3                                       0.000093   
4                                       0.000802   
...                                          ...   
3219                                    0.000000   
3220                                    0.000000   
3221                                    0.000000   
3222                                    0.000000   
3223                                    0.000000   

      golf_irrigation_total_energy_bbtu  \
0                              0.000298   
1                              0.008065   
2                              0.000875   
3                              0.000298   
4                              0.001360   
...                                 ...   
3219                           0.000000   
3220                           0.000000   
3221                           0.000000   
3222                           0.000000   
3223                           0.000000   

      golf_irrigation_total_rejected_energy_bbtu  \
0                                       0.000104   
1                                       0.002823   
2                                       0.000306   
3                                       0.000104   
4                                       0.000476   
...                                          ...   
3219                                    0.000000   
3220                                    0.000000   
3221                                    0.000000   
3222                                    0.000000   
3223                                    0.000000   

      golf_irrigation_total_energy_services_bbtu  livestock_total_energy_bbtu  \
0                                       0.000194                     0.000587   
1                                       0.005242                     0.000919   
2                                       0.000569                     0.001230   
3                                       0.000194                     0.000218   
4                                       0.000884                     0.002929   
...                                          ...                          ...   
3219                                    0.000000                     0.001842   
3220                                    0.000000                  

In [23]:
test_dict = {"Gfg" : [4, 7, 5]}
  
# printing original dictionary
print("The original dictionary is : " + str(test_dict))
  
# initializing K 
K = 1
  
# one liner, values() getting all value according to keys
res = [sub[K] for sub in test_dict.values()]
res

The original dictionary is : {'Gfg': [4, 7, 5]}


SyntaxError: can't use starred expression here (Temp/ipykernel_14748/3505006981.py, line 10)

In [3]:
df = flow.calc_energy_agriculture(data=data, total=True)
df
#df.groupby("State", as_index=False).sum()

TypeError: unhashable type: 'set'

In [6]:
df = flow.prep_pumping_intensity_data()
df

FIPS State             County  groundwater_pumping_bbtu_per_mg  \
0     01001    AL     Autauga County                         0.004775   
1     01003    AL     Baldwin County                         0.004775   
2     01005    AL     Barbour County                         0.004775   
3     01007    AL        Bibb County                         0.004775   
4     01009    AL      Blount County                         0.004775   
...     ...   ...                ...                              ...   
3218  72151    PR  Yabucoa Municipio                         0.004977   
3219  72153    PR    Yauco Municipio                         0.004977   
3220  78010    VI   St. Croix County                         0.004977   
3221  78020    VI    St. John County                         0.004977   
3222  78030    VI  St. Thomas County                         0.004977   

      surface_water_pumping_bbtu_per_mg  wastewater_pumping_bbtu_per_mg  
0                              0.001863                        0.001863  
1                              0.001863                        0.001863  
2                              0.001863                        0.001863  
3                              0.001863                        0.001863  
4                              0.001863                        0.001863  
...                                 ...                             ...  
3218                           0.001660                        0.001660  
3219                           0.001660                        0.001660  
3220                           0.001660                        0.001660  
3221                           0.001660                        0.001660  
3222                           0.001660                        0.001660  

[3223 rows x 6 columns]

In [8]:
df = flow.prep_irrigation_fuel_data()
df

FIPS State             County  electricity_pumping_pct  \
0     01001    AL     Autauga County                 0.470056   
1     01003    AL     Baldwin County                 0.470056   
2     01005    AL     Barbour County                 0.470056   
3     01007    AL        Bibb County                 0.470056   
4     01009    AL      Blount County                 0.470056   
...     ...   ...                ...                      ...   
3218  72151    PR  Yabucoa Municipio                 0.577682   
3219  72153    PR    Yauco Municipio                 0.577682   
3220  78010    VI   St. Croix County                 0.577682   
3221  78020    VI    St. John County                 0.577682   
3222  78030    VI  St. Thomas County                 0.577682   

      oil_pumping_pct  natural_gas_pumping_pct  
0            0.524731                 0.005213  
1            0.524731                 0.005213  
2            0.524731                 0.005213  
3            0.524731                 0.005213  
4            0.524731                 0.005213  
...               ...                      ...  
3218         0.346894                 0.075424  
3219         0.346894                 0.075424  
3220         0.346894                 0.075424  
3221         0.346894                 0.075424  
3222         0.346894                 0.075424  

[3223 rows x 6 columns]

In [8]:
df = flow.calc_electricity_wastewater(data=data)
df

FIPS State             County  electricity_wastewater_advanced_bbtu  \
0     01001    AL     Autauga County                               7532.00   
1     01003    AL     Baldwin County                              18830.00   
2     01005    AL     Barbour County                                  0.00   
3     01007    AL        Bibb County                                  0.00   
4     01009    AL      Blount County                               2399.48   
...     ...   ...                ...                                   ...   
3219  72151    PR  Yabucoa Municipio                                  0.00   
3220  72153    PR    Yauco Municipio                                  0.00   
3221  78010    VI   St. Croix County                                  0.00   
3222  78020    VI    St. John County                                  0.00   
3223  78030    VI  St. Thomas County                                  0.00   

      wastewater_advanced_rejected_energy_bbtu  \
0                                     2636.200   
1                                     6590.500   
2                                        0.000   
3                                        0.000   
4                                      839.818   
...                                        ...   
3219                                     0.000   
3220                                     0.000   
3221                                     0.000   
3222                                     0.000   
3223                                     0.000   

      wastewater_advanced_energy_services_bbtu  \
0                                     4895.800   
1                                    12239.500   
2                                        0.000   
3                                        0.000   
4                                     1559.662   
...                                        ...   
3219                                     0.000   
3220                                     0.000   
3221                                     0.000   
3222                                     0.000   
3223                                     0.000   

      electricity_wastewater_secondary_bbtu  \
0                                    137.28   
1                                  14988.48   
2                                   4447.04   
3                                   1526.72   
4                                    501.28   
...                                     ...   
3219                                2059.20   
3220                                3348.80   
3221                                9630.40   
3222                                   0.00   
3223                                   0.00   

      wastewater_secondary_rejected_energy_bbtu  \
0                                        48.048   
1                                      5245.968   
2                                      1556.464   
3                                       534.352   
4                                       175.448   
...                                         ...   
3219                                    720.720   
3220                                   1172.080   
3221                                   3370.640   
3222                                      0.000   
3223                                      0.000   

      wastewater_secondary_energy_services_bbtu  \
0                                        89.232   
1                                      9742.512   
2                                      2890.576   
3                                       992.368   
4                                       325.832   
...                                         ...   
3219                                   1338.480   
3220                                   2176.720   
3221                                   6259.760   
3222                                      0.000   
3223                                      0.000   

      electricity_wastewater_primary_bbtu  \
0                               

In [ ]:
df = flow.calc_sectoral_use_energy_discharge(data=data, total=True)
df

In [ ]:
data.to_csv(r'C:\Users\mong275\Local Files\Sankey_Notebook_Development\SankeyFiles\run4.csv')

To override the data inputs, flow.configure_data() can be run.
If run with no arguments provided, this function aggregates all baseline data to calculate flow values for the united states in 2015. Parameters are set, however, such that this function can be used to swap out datafiles. 

For example, if alternative US county level data was available for interbasin transfer electricity use, that could be used instead of the default data.

In [ ]:
df = flow.configure_data()
df

In [6]:
is_NaN = data.isnull()
row_has_NaN = is_NaN.any(axis=1)
rows_with_NaN = data[row_has_NaN]
rows_with_NaN

FIPS State                  County  population  \
3143  72001    PR      Adjuntas Municipio      18.579   
3144  72003    PR        Aguada Municipio      39.530   
3145  72005    PR     Aguadilla Municipio      55.804   
3146  72007    PR  Aguas Buenas Municipio      26.915   
3147  72009    PR      Aibonito Municipio      24.040   
...     ...   ...                     ...         ...   
3219  72151    PR       Yabucoa Municipio      35.082   
3220  72153    PR         Yauco Municipio      37.679   
3221  78010    VI        St. Croix County      49.250   
3222  78020    VI         St. John County       4.060   
3223  78030    VI       St. Thomas County      50.264   

      fresh_groundwater_pws_mgd  fresh_surface_water_pws_mgd  \
3143                       0.27                         1.80   
3144                       0.72                        21.76   
3145                       0.36                         3.47   
3146                       0.23                        13.92   
3147                       0.69                         1.66   
...                         ...                          ...   
3219                       1.57                         2.16   
3220                       0.00                         4.20   
3221                       0.77                         0.00   
3222                       0.01                         0.00   
3223                       0.13                         0.00   

      saline_groundwater_pws_mgd  saline_surface_water_pws_mgd  \
3143                         0.0                          0.00   
3144                         0.0                          0.00   
3145                         0.0                          0.00   
3146                         0.0                          0.00   
3147                         0.0                          0.00   
...                          ...                           ...   
3219                         0.0                          0.00   
3220                         0.0                          0.00   
3221                         0.0                          1.36   
3222                         0.0                          0.06   
3223                         0.0                          1.94   

      pws_residential_mgd  total_pws_mgd  ...  wind_commercial_bbtu  \
3143                 1.80           2.07  ...                   NaN   
3144                 5.43          22.48  ...                   NaN   
3145                 7.51           3.83  ...                   NaN   
3146                 2.40          14.15  ...                   NaN   
3147                 1.92           2.35  ...                   NaN   
...                   ...            ...  ...                   ...   
3219                 3.57           3.73  ...                   NaN   
3220                 3.18           4.20  ...                   NaN   
3221                 1.67           2.13  ...                   NaN   
3222                 0.03           0.07  ...                   NaN   
3223                 1.04           2.07  ...                   NaN   

      petroleum_production_bbtu  natural_gas_production_bbtu  \
3143                        0.0                          0.0   
3144                        0.0                          0.0   
3145                        0.0                          0.0   
3146                        0.0                          0.0   
3147                        0.0                          0.0   
...                         ...                          ...   
3219                        0.0                          0.0   
3220                        0.0                          0.0   
3221                        0.0                          0.0   
3222                        0.0                          0.0   
3223                        0.0                          0.0   

      refuse_coal_shortton  surface_coal_shortton  underground_coal_shortton  \
3143                   0.0                    0.0         

In [ ]:
df2.to_csv(r'C:\Users\mong275\Local Files\Sankey_Notebook_Development\SankeyFiles\run2.csv')

<h1> Public Water Supply </h1>

## Water in Public Water Supply

### Public Water Supply Withdrawals

The USGS 2015 dataset provides public water supply withdrawals directly under data codes 'PS-WGWFr', 'PS-WSWFr','PS-WGWSa', 'PS-WSWSa' for each county.

In [7]:
df = flow.prep_water_use_2015(variables=['State','County','PS-WGWFr', 'PS-WSWFr','PS-WGWSa', 'PS-WSWSa'])
df.head()               

KeyError: "['PS-WGWFr', 'PS-WSWFr', 'PS-WGWSa', 'PS-WSWSa'] not in index"

### Public Water Supply Deliveries to Sectors, Exports, and Imports

The USGS 2015 dataset provide public water supply data in three categories: (1) total public water supply withdrawals, (2) public water supply deliveries to domestic end users, and (3) public water supply deliveries to thermoelectric cooling.

The 1995 USGS dataset provides public water supply deliveries to domestic, industrial, commercial, thermoelectric cooling, and losses/exports separately. To estimate the deliveries to commercial and industrial customers in 2015, the ratio of each to the sum of deliveries to residential and thermoelectric cooling is calculated from the 1995 data. These ratios are applied to the 2015 residential and thermoelectric cooling public water supply delivery values. 

Additionally, public water suppy exports and imports are calculated. When the sum of water deliveries is greater than the water withdrawn for public water supply, the difference is considered to be imported. When water withdrawn is greater, excess water after deliveries is considered to be exported.

Caveats and assumptions:
* This methodology assumes that if no public water supply was delivered to residential or thermoelectric cooling in 1995, then deliveries to industrial and commercial sectors would likewise be zero.

Potential mods
 for import pws counties, take 15% of withdrawals and imports as losses and add it on to the top of imports
 for export pws counties, take 15% of withdrawals as losses, take it out of exports, any remaining put as imports

In [ ]:
df = flow.calc_pws_discharge()
df.head()

## Energy in Public Water Supply

Energy in public water supply includes
groundwater pumping, surface water pumping, desalination (when applicable), surface water treatment, groundwater treatment, distribution, and interbasin transfers (when appropriate).

An assumption is made that water that goes through desalination additionally requires surface water or groundwater treatment

The same groundwater and surface water pumping intensities are used from agriculture sector.

Note that interbasin transfer data that is provided to this function is already in bbtu, not in water flows. 
For non-ibt data, the option is provided for the user to use region-level energy intensity values (as in groundwater and surface water pumping). When values are not available, assumptions for each in kWh/MG are set as default values that can be overriden.

intensity values are used from tidwell and LLNL

In [ ]:
df = flow.calc_electricity_public_water_supply(total=True)
df.head()

# Wastewater Treatment
Wastewater treatment facility data is provided by EPA for the year 2012. These values are assumed for 2015. Values are provided on a treatment plant level with data on the following variables: 
Flow:
    * Total water flows (MGD) treated by the facility
    * Total infiltration flows treated by the facility (leaks into pipes and conduits)
Treatment:
    * Treatment type of each treatment plant (primary, secondary, advanced)
Discharge:
    * discharge location of each plant, broken down by percent of flow to each discharge location (e.g., ocean discharge, suface discharge, irrigation reuse)
Location:
    * county information is included for each power plant
    
This data allows us to determine total water flows processed and discharged by wastewater facilities at a county level by treatment type. The treatment type allows us to calculate the electricity use by county for wastewater treatment,

## Water in Wastewater Treatment

### Water Inflows to Wastewater Treatment Facilities
Total flows processed by wastewater treatment facilities are provided in the EPA dataset directly. However, this data will have discrepencies from the water data delivered to various sectors. To estimate water flows from residential, commercial, and industrial sectors to wastewater treatment facilities, an assumption is made that all water delivered from public water supply to these sectors is sent to wastewater treatment facilities. This assumption is also used to fill wastewater treatment flow values for regions that are unavailable in the EPA dataset. For example, south carolina does not have wastewater treatment facility values in the EPA dataset. It is assumed, therefore, that total wastewater treatment flows in south carolina and equal to the total public water supply flows for each county and is attributed to municipal discharge rather than infiltration. It is assumed that 80% of treatment flows are advanced treatment and 20% are secondary treatment. It is also assumed that all discharge is sent to surface. This follows average values for the state of North Carolina which were used as a proxy in this instance.

### Quantity of Wastewater Processed by Treatment Type
Wastewater treatment plant treatment type is provided in the EPA data directly. These values are binned for simplification into no treatment, primary, secondary, and advanced treatment. When no treatment type was provided for a facility, it was assumed that they were secondary treatment facilities.

To see the amount of water (MGD) processed within each county by treatment type, run the cell below.

In [ ]:
df = flow.prep_wastewater_data()
df.head()

### Wastewater Treatment Facility Discharge

Wastewater treatment plant discharge location is provided in the EPA data directly. These values are binned for simplification into the following categories:

* 'outfall to surface waters': 'ww_sd',
* 'ocean discharge': 'ww_od',
* 'deep well': 'ww_gd',
* "reuse: industrial": 'ww_in',
* 'evaporation': 'ww_evap',
* 'spray irrigation': 'ww_ir',
* 'overland flow no discharge': 'ww_ww',
* 'overland flow with discharge': 'ww_sd',
* 'discharge to another facility': 'ww_ww',
* 'combined sewer overflow (cso) discharge': 'ww_sd',
* 'other': 'ww_sd',
* 'discharge to groundwater': 'ww_gd',
* 'no discharge, unknown': 'ww_ww',
* 'reuse: irrigation': 'ww_ir',
* 'reuse: other non-potable': 'ww_sd',
* 'reuse: indirect potable': 'ww_sd',
* 'reuse: potable': 'ww_ps',
* 'reuse: groundwater recharge': 'ww_gd'

In [ ]:
df = flow.prep_wastewater_data()
df = df[['FIPS','ww_evap','ww_gd','ww_in','ww_ir','ww_od','ww_ps','ww_sd','ww_ww']]
df.head()

## Energy in Wastewater Treatment

Wastewater treatment facilities have various levels of treatment from primary to advanced. Each treatment level has an increasing level of energy intensity to process the water. Intensity value estimates for wastwater treatment were used from Pabi et al. [17].

# Agriculture

## Water in Agriculture

## Energy in Agriculture

Energy use in agriculture consists of groundwater pumping, surface water pumping, and interbasin transers. USDA's Farm and Ranch Irrigation Survey (FRIS) provides state-by-state data on irrigation groundwater depth and average irrigation pressurization levels for irrigation within a state, enabling the calculation of pump electricity consumption for both groundwater and surface water pumping [13]. The 2013 survey is the closest year available to 2015 values. It is assumed that values do not vary significantly between the two years.

The methodology for calculating groundwater and surface water pumping energy is described in Tidwell et al [14]. However, where that publication uses well depth to water to calculate total differential height, the total well depth is used here as a way to offset some of the losses due to friction that would occur in the piping, as described in Lawrence Berkeley National Laboratory (LBNL) Home Energy Saver & Score: Engineering Documentation [15].

Pump efficiency is assumed to be the average (46.5%) of the range (34-59%) listed in [14]. State-level intensity rates are calculated here and applied to the county level water in agriculture values. The general methodology implemented here to determine total energy in agriculture is adapted from [5], not including the additions previously mentioned.

### Groundwater & Surface Water Pumping

### Interbasin Transfers

# Electricity Generation

## Energy in Electricity Generation

US Energy Information Administration (EIA) 923 data for 2015 is used to determine electricity generation by generator type within each interconnect [12]. The datafile contains information on generation at the plant-level. EIA additionally provides data on which North American Electric Reliability Corporation (NERC) balancing region each plant belongs to. By mapping each NERC region to its associated interconnect, generation within each interconnect can be determined directly.

From EIA, the AER Fuel Type Code is used to determine generator type. EIA provides generator type information at a high level of detail. For this reason different generator types are grouped for simplicity using a generator key (GenType923Key.xlsx). For example, different types of natural gas generators are all grouped into "Natural Gas" as a type.

Energy production is provided as "Net Generation (Megawatthours)" in the EIA dataset, "Total Fuel Consumption (MMBTU)" is the MMBTU equivalent of fuel going to each plant for electricity generation. The difference between these two values indicates the amount of energy generation that does not go to energy losses.

## will be filled in with calculate function

## Water in Electricity Generation

# End Use Sectors

## Residential Sector

### Water in Residential Sector

### Energy in Residential Sector

#### Energy (electricity) Demand in Residential Sector

##### weighted by population. <b> consider splitting out by usage if data is available

In [ ]:
df = flow.prep_electricity_demand_data()
df

#### Energy (fuel) Demand in Residential Sector

## Commercial Sector

### Water in Commercial Sector

### Energy in Commercial Sector

## Industrial Sector

### Water in Industrial Sector

### Energy in Industrial Sector

## Transportation Sector

### Energy in Energy Sector

# Energy Production

## Biomass

### Energy in Biomass Production (Ethanol)

To find the ethanol production per county, this calculation takes 2015 eia data on ethanol plant capacity which has added locational data by each plant and combines with state level biomass (ethanol) production from prep_state_fuel_production_data() to split out state total by county. Returns a dataframe of ethanol production (bbtu) by county FIPS for each county in the US for 2015.

ethanol plant production filler for wyoming (in state total but not location data) ,Torrington, WY100%, https://neo.ne.gov/programs/stats/122/2015/122_201512.htm

#### <b> <ul> <font color='red'> Consider looking for estimates for electricity use in biomass production to split out from industrial

### Water in Biomass Production (Ethanol)

#### Water used in irrigation of corn used for ethanol

Data on water used in the irrigation of corn ultimately used for ethanol production does not exist explicitly at any level of granularity and must be calculated out from larger state-level corn growth irrigation intensity estimates.

The percent of corn grown for ethanol in the US was used from US DOE Alternative Fuels Data Center [21]. Total corn production for ethanol (5.22 billion bushels) was divided by total corn producton (13.6 billion bushels) for 2015 and calculated to be 38.406%. Data on a state or county-level does not exist, therefore, this percentage is applied to all states with corn production. A state-level fraction of corn used for ethanol production could be calculated from the amount of ethanol produced in a state and applying an assumption of corn required per unit of ethanol, however, not all corn used in ethanol is produced in the state that uses it which lead to states such as CA with an ethanol fraction of corn growth of nearly 900% of their total corn production. Given that we are unable to determine which states the imported corn for ethanol would be coming from, the decision was made to apply the US total fraction to all states.

Total water used in the production of corn for ethanol is therefore the assumed 38.406% multiplied by the product of the total amount of irrigated acres of corn (AC_Irr_Corn) and the calculated irrigation intensity for that state (total acres applied to all crops / total acres irrigated, all crops) provided from USDA Census of Agriculture Irrigation and Water Management Survey [22], Table 35 and Table 4, respectively. The irrigation intensity for corn is calculated from the average irrigation intensity for the state.

To split the total MGY applied between fresh surface and groundwater, USDA FRIS Census of Agriculture data on the fraction of surface water (surface water + off farm water) from total water supply (including groundwater) in corn production is used from Table 37 of [22]. Off-farm sources are assumed to be surface water.

Some states that grow corn do not provide data on the water source breakdown (surface vs. ground) for their irrigation. The surface to groundwater ratio for all irrigation in a state is substituted for these states. These values are calculated from the USGS 2015 water dataset.

In [ ]:
df = flow.prep_county_water_corn_biomass_data()

#### Water used in the production of ethanol (fermentation)

#### <b> <ul> <font color='red'> Consider looking for estimates for water used in the fermentation process of ethanol production, not just corn growth.

## Petroleum

### Energy in Petroleum Production (Conventional and Unconventional Oil Drilling)

 crude oil production by county fillers
Idaho: https://ogcc.idaho.gov/wp-content/uploads/sites/3/2017/06/2016-10-6-state-of-idaho-releases-oil-gas-production-data.pdf
100% of Idaho petroleum and natural gas is from Payette County

Alaska,
https://www.commerce.alaska.gov/web/aogcc/Data.aspx
almost all oil is produced along the north slope. Calculated from oil production download. The rest is on the kenai peninsula (cook inlet basin)

Maryland
https://www.eia.gov/state/analysis.php?sid=MD#34
assumed 50/50 split between Garret and Allegany Counties

Nevada
https://gisweb.unr.edu/OilGas/
Assumed to be in nye county with producing oil wells (only 4 btu)

Oregon
https://www.oregongeology.org/mlrr/oilgas-report.htm
Only one county (columbia county)

### Water in Petroleum Production (Conventional and Unconventional Oil Drilling)

## Natural Gas

### Energy in Natural Gas Production (Unconventional Drilling)

### Water in Natural Gas Production (Unconventional Drilling)

## Coal

### Energy in Coal Production (Coal Mining)

### Water in Coal Production (Coal Mining)

# Intensities & Percentages

#### Conveyance Losses as a Percent of Water Withdrawn for Irrigation
Conveyance loss fraction for irrigation is calculated from the USGS 1995 water use dataset. It is calculated as the total conveyance losses (MGD) divided by the total water withdrawn (MGD) for irrigation. Looking at the fractions at the county level shows that a small subset of counties have unreliable values where conveyance losses are far greater than the amount withdrawn for irrigation. As a result, the ability to cap the loss fraction at a specified amount it provided and replace it with the average of the specified region. The default loss cap fraction is .90.

In [ ]:
df = flow.prep_water_use_1995()
df = flow.calc_conveyance_loss_frac(df)
df.head()

##### Change in loss cap

In [ ]:
df = flow.prep_water_use_1995()
df = flow.calc_conveyance_loss_frac(df, loss_cap_amt=.20)
df.head()

In [ ]:
help(flow.plot_bar)

In [ ]:
flow.plot_bar(df, "State","IR_CLoss_Frac", "State", y_axis_title = "Irrigation Conveyance Loss Fraction", x_axis_title="State")

#### Water Use Intensity for Hydroelectric Generation

In [ ]:
df = flow.calc_hydroelectric_water_intensity(intensity_cap=True, intensity_cap_amt=165, region_avg=False)

df.head()

In [ ]:
flow.plot_bar(df, "State","HY_IF", "State")